src/VisualStudio/Core/Def/Implementation/Remote/ServiceHubRemoteHostClient.cs

```diff
@@ -46,77 +47,97 @@ private enum GlobalNotificationState
 
-                    client = new ServiceHubRemoteHostClient(workspace, primary.Logger, connectionManager, remoteHostStream);
+                var client = new ServiceHubRemoteHostClient(workspace, hubClient.Logger, connectionManager, remoteHostStream);
 
-                    var uiCultureLCID = CultureInfo.CurrentUICulture.LCID;
-                    var cultureLCID = CultureInfo.CurrentCulture.LCID;
+                var uiCultureLCID = CultureInfo.CurrentUICulture.LCID;
+                var cultureLCID = CultureInfo.CurrentCulture.LCID;
 
+                bool success = false;
+                try
+                {
                     // make sure connection is done right
-                    var host = await client._rpc.InvokeWithCancellationAsync<string>(
+                    _ = await client._rpc.InvokeWithCancellationAsync<string>(
                         nameof(IRemoteHostService.Connect), new object[] { clientId, uiCultureLCID, cultureLCID, TelemetryService.DefaultSession.SerializeSettings() }, cancellationToken).ConfigureAwait(false);
 
-                    client.Started();
-
-                    return client;
+                    success = true;
                 }
-                catch (ConnectionLostException ex)
+                catch (Exception e)
                 {
-                    RemoteHostCrashInfoBar.ShowInfoBar(workspace, ex);
-
-                    Shutdown(ex);
-
-                    // dont crash VS because OOP is failed to start. we will show info bar telling users to restart
-                    // but never physically crash VS.
-                    return null;
-                }
-                catch (SoftCrashException ex)
-                {
-                    Shutdown(ex);
+                    // translate to our own cancellation if it is raised.
+                    cancellationToken.ThrowIfCancellationRequested();
 
-                    // at this point, we should have shown info bar (RemoteHostCrashInfoBar.ShowInfoBar) to users
-                    // returning null here will disable OOP for this VS session. 
-                    // * Note * this is not trying to recover the exception. but giving users to time
-                    // to clean up before restart VS
-                    return null;
+                    throw SoftCrash(workspace, hubClient, e, "Failed to connect to ServiceHub", cancellationToken);
                 }
-                catch (Exception ex)
+                finally
                 {
-                    Shutdown(ex);
-                    throw;
+                    if (!success)
+                    {
+                        client.Dispose();
+                    }
                 }
 
-                void Shutdown(Exception ex)
-                {
-                    // make sure we shutdown client if initializing client has failed.
-                    client?.Shutdown();
+                client.Started();
```

Comment:
is this safe to call .Started after .Dispose may have been called?  Can we add comments to the effect that this is ok?

Reply:
oh.  nevermind.  the 'throw' will prevent us from geting hre.  LGTM>